### ISSUE TITLE
zjobs: high cpu usage

### ISSUE THREAD

**Issue by Jack-Ji** [michal-z/zig-gamedev#175]
I'm trying to parallel frontend of my software rasterizer using zjobs. The default JobQueue occupied at least 80% cpu, even when I'm not scheduling any jobs. 

OS: windows 11 x64
zig: latest master

**Comment by michal-z** [comment]
@garettbass is this known issue?

**Comment by hazeycode** [comment]
zjobs has threads busy wait using sleep. You can configure the sleep time for your JobQueue by changing `idle_sleep_ns` form the default of 10 ns. Note this is probably below the accuracy error of the OS's scheduler. But by making it bigger you should see less CPU usage. By doing this you are trading "wake up" performance i.e. how long it can take a job to get picked up by a thread. You basically get to pick between wasting energy or increasing latency of jobs starting. You can get better by not using sleep and instead suspending and signalling threads (but it's kinda complicated to do cross-platform).
cyptocode write a nice blog post, showing how this looks at a high level using the async/await and standard event loop abstractions: https://dev.to/stein/async-cpu-bound-workers-in-zig-24b0

**Comment by Jack-Ji** [comment]
I've thought of adjusting sleeping interval actually. However, I also   think the default configuration should works without obviously problem across platforms.

**Comment by hazeycode** [comment]
> I've thought of adjusting sleeping interval actually. However, I also think the default configuration should works without obviously problem across platforms.

10 ns is really tiny! I'd guess that means basically that the thread is suspended and immediately woken by the OS scheduler.

**Comment by Jack-Ji** [comment]
> > I've thought of adjusting sleeping interval actually. However, I also think the default configuration should works without obviously problem across platforms.
> 
> 
> 
> 10 ns is really tiny! I'd guess that means basically that the thread is suspended and immediately woken by the OS scheduler.

Yeah, I agree with you. May be we can change it to some more reasonable value? Any way, I'll try it later to see what happens.

**Comment by hazeycode** [comment]
> > > I've thought of adjusting sleeping interval actually. However, I also think the default configuration should works without obviously problem across platforms.
> > 
> > 
> > 10 ns is really tiny! I'd guess that means basically that the thread is suspended and immediately woken by the OS scheduler.
> 
> Yeah, I agree with you. May be we can change it to some more reasonable value? Any way, I'll try it later to see what happens.

The problem is it's difficult to pick a number that is going to work well everywhere. You can provide a different value per target OS but you will also have to tune it per use case in your program.

That said, 10 ns I don't think is a reasonable value for any OS scheduler.

**Comment by garettbass** [comment]
This is a known issue.  I have not yet tried to address it, but I have some ideas how I might.

Please consider that I have written the simplest possible implementation I could implement reliably.  Pull requests from devs more experienced with this problem space are quite welcome.

**Comment by michal-z** [comment]
@garettbass you already did great job with job system :) All the improvements are of course awesome and very welcome but don't worry if you don't have time for them. We will do it.

I will be working on a software fractal renderer as a showcase for `zjobs` and `zmath` - this will give me an opportunity to delve into `zjobs` sources.

**Comment by Avokadoen** [comment]
This issue relates to #130
